src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/AbstractAddItemCommandHandler.cs

```diff
@@ -0,0 +1,122 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Input;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Input.Commands
+{
+    /// <summary>
+    /// Provides support for all Add Item commands that operate on <see cref="IProjectTree"/> nodes, across C# and VB
+    /// </summary>
+    internal abstract partial class AbstractAddItemCommandHandler : IAsyncCommandGroupHandler
+    {
+        protected static readonly Guid LegacyCSharpPackageGuid = new Guid("{FAE04EC1-301F-11d3-BF4B-00C04F79EFBC}");
+        protected static readonly Guid LegacyVBPackageGuid = new Guid("{164B10B9-B200-11d0-8C61-00A0C91E29D5}");
+        private readonly ConfiguredProject _configuredProject;
+        private readonly IPhysicalProjectTree _projectTree;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IVsUIService<IVsAddProjectItemDlg> _addItemDialog;
+        private readonly IVsUIService<IVsShell> _vsShell;
+        private readonly Lazy<Dictionary<long, List<TemplateDetails>>> _commandMap;
```

Comment:
These never change, I would type them as immutable types to prevent them from ever being changed.